examples/image_viewporter.rs

```diff
@@ -0,0 +1,327 @@
+                    viewer.window.wl_surface().commit();
+                }
+            }
+
+            // Initiate the first draw.
+            viewer.first_configure = false;
+        }
+        self.draw(conn, qh);
+    }
+}
+
+impl ShmHandler for State {
+    fn shm_state(&mut self) -> &mut ShmState {
+        &mut self.shm_state
+    }
+}
+
+impl State {
+    pub fn draw(&mut self, _conn: &Connection, _qh: &QueueHandle<Self>) {
+        for viewer in &mut self.windows {
+            if viewer.first_configure || !viewer.damaged {
+                continue;
+            }
+            let window = &viewer.window;
+            let width = viewer.image.width();
+            let height = viewer.image.height();
+            let stride = width as i32 * 4;
+            let pool = self.pool.as_mut().unwrap();
+
+            let (buffer, canvas) = pool
+                .create_buffer(width as i32, height as i32, stride, wl_shm::Format::Argb8888)
+                .expect("create buffer");
+
+            // Draw to the window
+            for (pixel, argb) in viewer.image.pixels().zip(canvas.chunks_exact_mut(4)) {
+                // We do this in an horribly inefficient manner, for the sake of simplicity.
+                // We'll send pixels to the server in ARGB8888 format (this is one of the only
+                // formats that are guaranteed to be supported), but image provides it in
+                // big-endian RGBA8888, so we need to do the conversion.
+                argb[3] = pixel.0[3];
+                argb[2] = pixel.0[0];
+                argb[1] = pixel.0[1];
+                argb[0] = pixel.0[2];
+            }
+
+            // Damage the entire window (using the real dimensions)
+            window.wl_surface().damage_buffer(0, 0, viewer.width as i32, viewer.height as i32);
+            viewer.damaged = false;
+
+            // Set the entire buffer as the source area for the viewport.
+            // Destination was set during configure.
+            viewer.viewport.set_source(0.0, 0.0, viewer.width as f64, viewer.height as f64);
+
+            // Attach and commit to present.
+            buffer.attach_to(window.wl_surface()).expect("buffer attach");
+            window.wl_surface().commit();
+        }
+    }
+}
+
+delegate_compositor!(State);
+delegate_output!(State);
+delegate_shm!(State);
+
+delegate_xdg_shell!(State);
+delegate_xdg_window!(State);
+
+delegate_simple!(State, WpViewporter, 1);
+
+delegate_registry!(State);
+
+impl ProvidesRegistryState for State {
+    fn registry(&mut self) -> &mut RegistryState {
+        &mut self.registry_state
+    }
+
+    registry_handlers!(CompositorState, OutputState, ShmState, XdgShellState, XdgWindowState, SimpleGlobal<WpViewporter, 1>);
+}
+
+impl AsMut<SimpleGlobal<WpViewporter, 1>> for State {
+    fn as_mut(&mut self) -> &mut SimpleGlobal<WpViewporter, 1> {
+        &mut self.viewporter
+    }
+}
+
+impl Dispatch<WpViewport, ()> for State {
```

Comment:
Is there not a delegate macro simple global provides for this?

Reply:
This is for `WpViewport`; the delegate is use (and can only be used) for the `WpViewporter` (which is the global).